File: 00_Alternate_Languages/45_Hello/java/Hello.java
import java.util.Scanner;

/**
 * Game of Hello
 * <p>
 * Based on the BASIC game of Hello here
 * https://github.com/coding-horror/basic-computer-games/blob/main/45%20Hello/hello.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's BASIC game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 *
 * Converted from BASIC to Java by Darren Cardenas.
 */

public class Hello {

  private static final int MONEY_WAIT_MS = 3000;

  private final boolean goodEnding = false;

  private final Scanner scan;  // For user input

  public Hello() {

    scan = new Scanner(System.in);

  }  // End of constructor Hello

  public void play() {

    showIntro();
    startGame();

  }  // End of method play

  private static void showIntro() {

    System.out.println(" ".repeat(32) + "HELLO");
    System.out.println(" ".repeat(14) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
    System.out.println("\n\n");

  }  // End of method showIntro

  private void startGame() {

    boolean moreProblems = true;

    String userCategory = "";
    String userName = "";
    String userResponse = "";

    // Name question
    System.out.println("HELLO.  MY NAME IS CREATIVE COMPUTER.\n\n");
    System.out.print("WHAT'S YOUR NAME? ");
    userName = scan.nextLine();
    System.out.println("");

    // Enjoyment question
    System.out.print("HI THERE, " + userName + ", ARE YOU ENJOYING YOURSELF HERE? ");

    while (true) {
      userResponse = scan.nextLine();
      System.out.println("");

      if (userResponse.toUpperCase().equals("YES")) {
        System.out.println("I'M GLAD TO HEAR THAT, " + userName + ".\n");
        break;
      }
      else if (userResponse.toUpperCase().equals("NO")) {
        System.out.println("OH, I'M SORRY TO HEAR THAT, " + userName + ". MAYBE WE CAN");
        System.out.println("BRIGHTEN UP YOUR VISIT A BIT.");
        break;
      }
      else {
        System.out.println(userName + ", I DON'T UNDERSTAND YOUR ANSWER OF '" + userResponse + "'.");
        System.out.print("PLEASE ANSWER 'YES' OR 'NO'.  DO YOU LIKE IT HERE? ");
      }
    }

    // Category question
    System.out.println("");
    System.out.println("SAY, " + userName + ", I CAN SOLVE ALL KINDS OF PROBLEMS EXCEPT");
    System.out.println("THOSE DEALING WITH GREECE.  WHAT KIND OF PROBLEMS DO");
    System.out.print("YOU HAVE (ANSWER SEX, HEALTH, MONEY, OR JOB)? ");

    while (moreProblems) {
      userCategory = scan.nextLine();
      System.out.println("");

      // Sex advice
      if (userCategory.toUpperCase().equals("SEX")) {
        System.out.print("IS YOUR PROBLEM TOO MUCH OR TOO LITTLE? ");
        userResponse = scan.nextLine();
        System.out.println("");

        while (true) {
          if (userResponse.toUpperCase().equals("TOO MUCH")) {
            System.out.println("YOU CALL THAT A PROBLEM?!!  I SHOULD HAVE SUCH PROBLEMS!");
            System.out.println("IF IT BOTHERS YOU, " + userName + ", TAKE A COLD SHOWER.");
            break;
          }
          else if (userResponse.toUpperCase().equals("TOO LITTLE")) {
            System.out.println("WHY ARE YOU HERE IN SUFFERN, " + userName + "?  YOU SHOULD BE");
            System.out.println("IN TOKYO OR NEW YORK OR AMSTERDAM OR SOMEPLACE WITH SOME");
            System.out.println("REAL ACTION.");
            break;
          }
          else {
            System.out.println("DON'T GET ALL SHOOK, " + userName + ", JUST ANSWER THE QUESTION");
            System.out.print("WITH 'TOO MUCH' OR 'TOO LITTLE'.  WHICH IS IT? ");
            userResponse = scan.nextLine();
          }
        }
      }
      // Health advice
      else if (userCategory.toUpperCase().equals("HEALTH")) {
        System.out.println("MY ADVICE TO YOU " + userName + " IS:");
        System.out.println("     1.  TAKE TWO ASPRIN");
        System.out.println("     2.  DRINK PLENTY OF FLUIDS (ORANGE JUICE, NOT BEER!)");
        System.out.println("     3.  GO TO BED (ALONE)");
      }
      // Money advice
      else if (userCategory.toUpperCase().equals("MONEY")) {
        System.out.println("SORRY, " + userName + ", I'M BROKE TOO.  WHY DON'T YOU SELL");
        System.out.println("ENCYCLOPEADIAS OR MARRY SOMEONE RICH OR STOP EATING");
        System.out.println("SO YOU WON'T NEED SO MUCH MONEY?");
      }
      // Job advice
      else if (userCategory.toUpperCase().equals("JOB")) {
        System.out.println("I CAN SYMPATHIZE WITH YOU " + userName + ".  I HAVE TO WORK");
        System.out.println("VERY LONG HOURS FOR NO PAY -- AND SOME OF MY BOSSES");
        System.out.println("REALLY BEAT ON MY KEYBOARD.  MY ADVICE TO YOU, " + userName + ",");
        System.out.println("IS TO OPEN A RETAIL COMPUTER STORE.  IT'S GREAT FUN.");
      }
      else {
        System.out.println("OH, " + userName + ", YOUR ANSWER OF " + userCategory + " IS GREEK TO ME.");
      }

      // More problems question
      while (true) {
        System.out.println("");
        System.out.print("ANY MORE PROBLEMS YOU WANT SOLVED, " + userName + "? ");
        userResponse = scan.nextLine();
        System.out.println("");

        if (userResponse.toUpperCase().equals("YES")) {
          System.out.print("WHAT KIND (SEX, MONEY, HEALTH, JOB)? ");
          break;
        }
        else if (userResponse.toUpperCase().equals("NO")) {
          moreProblems = false;
          break;
        }
        else {
          System.out.println("JUST A SIMPLE 'YES' OR 'NO' PLEASE, " + userName + ".");
        }
      }
    }

    // Payment question
    System.out.println("");
    System.out.println("THAT WILL BE $5.00 FOR THE ADVICE, " + userName + ".");
    System.out.println("PLEASE LEAVE THE MONEY ON THE TERMINAL.");

    // Pause
    try {
      Thread.sleep(MONEY_WAIT_MS);
    } catch (Exception e) {
      System.out.println("Caught Exception: " + e.getMessage());
    }

    System.out.println("\n\n");

    while (true) {
      System.out.print("DID YOU LEAVE THE MONEY? ");
      userResponse = scan.nextLine();
      System.out.println("");

      if (userResponse.toUpperCase().equals("YES")) {
        System.out.println("HEY, " + userName + "??? YOU LEFT NO MONEY AT ALL!");
        System.out.println("YOU ARE CHEATING ME OUT OF MY HARD-EARNED LIVING.");
        System.out.println("");
        System.out.println("WHAT A RIP OFF, " + userName + "!!!\n");
        break;
      }
      else if (userResponse.toUpperCase().equals("NO")) {
        System.out.println("THAT'S HONEST, " + userName + ", BUT HOW DO YOU EXPECT");
        System.out.println("ME TO GO ON WITH MY PSYCHOLOGY STUDIES IF MY PATIENTS");
        System.out.println("DON'T PAY THEIR BILLS?");
        break;
      }
      else {
        System.out.println("YOUR ANSWER OF '" + userResponse + "' CONFUSES ME, " + userName + ".");
        System.out.println("PLEASE RESPOND WITH 'YES' OR 'NO'.");
      }
    }

    // Legacy included unreachable code
    if (goodEnding) {
      System.out.println("NICE MEETING YOU, " + userName + ", HAVE A NICE DAY.");
    }
    else {
      System.out.println("");
      System.out.println("TAKE A WALK, " + userName + ".\n");
    }

  }  // End of method startGame

  public static void main(String[] args) {

    Hello hello = new Hello();
    hello.play();

  }  // End of method main

}  // End of class Hello

File: 00_Alternate_Languages/45_Hello/Swift/hello.swift

func tab(_ number_Of_Spaces: Int) {
    var spaces = ""
    
    for _ in 1...number_Of_Spaces {
        spaces += " "
    }
    print(spaces, terminator:"")
}


func get_Input() -> String {
    let input = readLine()
    return (input == nil ? "" : input!.uppercased())
}


func main()
{
    var done = false,
        answered = false,
        maybe_More = false,
        paid = false
    var reply = ""
    var name = "STRANGER"
    
    tab (33)
    print("HELLO")
    tab (15)
    print("CREATIVE COMPUTING   MORRISTOWN, NEW JERSEY\n")
    
    print("HELLO.  MY NAME IS CREATIVE COMPUTER.")
    print("WHAT'S YOUR NAME? ")
    let input = readLine()
    if (input != nil && input != "") {
        name = input!.uppercased()
    }
    
    print("\nHI THERE, \(name), ARE YOU ENJOYING YOURSELF HERE?")

    reply = get_Input()
    while (reply != "YES" && reply != "NO") {
        print("\(name), I DON'T UNDERSTAND YOUR ANSWER OF '\(reply)'.")
        print("PLEASE ANSWER 'YES' OR 'NO'. DO YOU LIKE IT HERE?")
        reply = get_Input()
    }
    
    if (reply == "YES") {
        print("\nI'M GLAD TO HEAR THAT, \(name).\n")
    }
    else {
        print("\nOH, I'M SORRY TO HEAR THAT, \(name). MAYBE WE CAN "
            + "BRIGHTEN UP YOUR VISIT A BIT.\n")
    }

    print("SAY, \(name), I CAN SOLVE ALL KINDS OF PROBLEMS EXCEPT "
        + "THOSE DEALING WITH GREECE.  WHAT KIND OF PROBLEMS DO "
        + "YOU HAVE (ANSWER SEX, HEALTH, MONEY, OR JOB)?")

    while (!done) {
        reply = get_Input()
        
        if (reply == "JOB") {
            print("\nI CAN SYMPATHIZE WITH YOU \(name).   I HAVE TO WORK "
                + "VERY LONG HOURS FOR NO PAY -- AND SOME OF MY BOSSES "
                + "REALLY BEAT ON MY KEYBOARD.    MY ADVICE TO YOU, \(name), IS TO "
                + "OPEN A RETAIL COMPUTER STORE.  IT'S GREAT FUN.\n")
        }

        else if (reply == "MONEY") {
            print("\nSORRY, \(name), I'M BROKE TOO.  WHY DON'T YOU SELL "
                + "ENCYCLOPEADIAS OR MARRY SOMEONE RICH OR STOP EATING "
                + "SO YOU WON'T NEED SO MUCH MONEY?\n")
        }
        
        else if (reply == "HEALTH") {
            print("\nMY ADVICE TO YOU \(name) IS:")
            print("         1.  TAKE TWO ASPRIN")
            print("         2.  DRINK PLENTY OF FLUIDS (ORANGE JUICE, NOT BEER!)")
            print("         3.  GO TO BED (ALONE)\n")
        }
        
        else if (reply == "SEX") {
            print("\nIS YOUR PROBLEM TOO MUCH OR TOO LITTLE?")
            
            answered = false
            while (!answered) {
                reply = get_Input()
                if (reply == "TOO MUCH") {
                    print("\nYOU CALL THAT A PROBLEM?!!  I SHOULD HAVE SUCH PROBLEMS!")
                    print("IF IT BOTHERS YOU, \(name), TAKE A COLD SHOWER.\n")
                    answered = true
                }
                else if (reply == "TOO LITTLE") {
                    print("\nWHY ARE YOU HERE IN SUFFERN, \(name)? YOU SHOULD BE "
                        + "IN TOKYO OR NEW YORK OR AMSTERDAM OR SOMEPLACE WITH SOME "
                        + "REAL ACTION.\n")
                    answered = true
                }
                else {
                    print("\nDON'T GET ALL SHOOK, \(name), JUST ANSWER THE QUESTION "
                        + "WITH 'TOO MUCH' OR 'TOO LITTLE'. WHICH IS IT?")
                }
            }
        }
        
        else {  // not one of the prescribed categories
            print("\nOH, \(name), YOUR ANSWER OF '\(reply)' IS GREEK TO ME.\n")
        }
        
        print("\nANY MORE PROBLEMS YOU WANT SOLVED, \(name)? ")
        
        maybe_More = true
        while (maybe_More) {
            reply = get_Input()
            if (reply == "NO") {
                done = true
                maybe_More = false
            }
            else if (reply == "YES") {
                print("\nWHAT KIND (SEX, MONEY, HEALTH, JOB) ")
                maybe_More = false
            }
            else {
                print("\nJUST A SIMPLE 'YES' OR 'NO' PLEASE, \(name). ")
            }
        } // no further questions
    } // end of 'not done' loop
    
    print("\nTHAT WILL BE $5.00 FOR THE ADVICE, \(name).")
    print("PLEASE LEAVE THE MONEY ON THE TERMINAL.")
    // pause a few seconds
    print("\n\n\nDID YOU LEAVE THE MONEY? ")
    reply = get_Input()
    while (!paid) {
        if (reply == "YES") {
            print("\nHEY, \(name)??? YOU LEFT NO MONEY AT ALL!")
            print("YOU ARE CHEATING ME OUT OF MY HARD-EARNED LIVING.\n")
            print("WHAT A RIP OFF, \(name)!!!\n")
            print("TAKE A WALK, \(name).")
            paid = true
        }
        else if (reply == "NO") {
            print("THAT'S HONEST, \(name), BUT HOW DO YOU EXPECT "
                + "ME TO GO ON WITH MY PSYCHOLOGY STUDIES IF MY PATIENTS "
                + "DON'T PAY THEIR BILLS?\n")
            print("NICE MEETING YOU, \(name), HAVE A NICE DAY.")
                paid = true
        }
        else {
            print("YOUR ANSWER OF '\(reply)' CONFUSES ME, \(name).")
            print("PLEASE RESPOND WITH 'YES' OR 'NO'.")
        }
    }
}

main()
